### PR TITLE
Don't define _LIBCPP_ENABLE_ASSERTIONS for clang++-19 and later

### PR DESCRIPTION
**Pull request by masterleinad**
`clang++-19` warns if we set `_LIBCPP_ENABLE_ASSERTIONS` and we already have the replacement in place.

### CODE REVIEW

File: cmake/setup_compiler_flags_gnu.cmake [hunk 1]
```diff
@@ -192,7 +192,10 @@ if (CMAKE_BUILD_TYPE MATCHES "Debug")
   # Enable invalid element access and other checks in the c++ standard libray:
   list(APPEND DEAL_II_DEFINITIONS_DEBUG "_GLIBCXX_ASSERTIONS")
   if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    list(APPEND DEAL_II_DEFINITIONS_DEBUG "_LIBCPP_ENABLE_ASSERTIONS")
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 19)
+      # _LIBCPP_ENABLE_ASSERTIONS was deprecated in clang++-19
+      list(APPEND DEAL_II_DEFINITIONS_DEBUG "_LIBCPP_ENABLE_ASSERTIONS")
+    endif()
```
Comment (bangerth) [inline cmake/setup_compiler_flags_gnu.cmake:198]: Is there any kind of replacement for newer clang versions? Surely they haven't completely gotten rid of their debug infrastructure.

Comment (masterleinad) [reply cmake/setup_compiler_flags_gnu.cmake:198]: Yes, `LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE` which we are already using below.

**Review by drwells — APPROVED** [review]

File: cmake/setup_compiler_flags_gnu.cmake (hunk 1 above)
Comment (bangerth) [reply cmake/setup_compiler_flags_gnu.cmake:198]: I see. Perhaps mention that in the comment you're adding here?

Comment (masterleinad) [reply cmake/setup_compiler_flags_gnu.cmake]: Here you go!

**Review by tamiko — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Comment by bangerth** [comment]
Thank you!

**Pull request merged**